compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -1567,6 +1573,92 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+            )) => {
+                let Some(hir::Node::Pat(hir::Pat { hir_id, .. })) = self.tcx.hir().find(*binding)
+                else {
+                    return expr;
+                };
+                let Some(parent) = self.tcx.hir().find(self.tcx.hir().parent_id(*hir_id)) else {
+                    return expr;
+                };
+
+                match parent {
+                    // foo.clone()
+                    hir::Node::Local(hir::Local { init: Some(init), .. }) => {
+                        self.note_type_is_not_clone_inner_expr(init)
+                    }
+                    // When `expr` is more complex like a tuple
+                    hir::Node::Pat(hir::Pat { hir_id: pat_hir_id, kind: hir::PatKind::Tuple(pats, ..), .. }) => {
+                        let Some(hir::Node::Local(hir::Local { init: Some(init), .. })) =
+                                self.tcx.hir().find(self.tcx.hir().parent_id(*pat_hir_id)) else {
+                            return expr;
+                        };
+
+                        let find_init = |init_tup: &'b [Expr<'b>]| {
+                            pats.iter().enumerate().filter(|x| x.1.hir_id == *hir_id).map(|(i,_)| init_tup.get(i).unwrap()).next()
+                        };
+
+                        match init.kind {
+                            ExprKind::Tup(init_tup) => {
+                                if let Some(init) = find_init(init_tup) {
+                                    self.note_type_is_not_clone_inner_expr(init)
+                                } else {
+                                    expr
+                                }
+                            }
+                            ExprKind::Block(_, _) => {
+                                let block_expr = init.peel_blocks();
+                                if let Expr { kind: ExprKind::Tup(init_tup), .. } = block_expr
+                                    && let Some(init) = find_init(init_tup)
+                                {
+                                    self.note_type_is_not_clone_inner_expr(init)
+                                } else {
+                                    expr
+                                }
+                            }
+                            _ => expr
+                        }
+                    }
+                    _ => expr
+                }
+            }
+            // If we're a closure recurse back into that closure
+            hir::ExprKind::Closure(hir::Closure { body: body_id, .. }) => {
+                let hir::Body { value: body_expr, .. } = self.tcx.hir().body(*body_id);
+                self.note_type_is_not_clone_inner_expr(body_expr)
+            }
+            // Similarly recurse into a block
+            hir::ExprKind::Block(hir::Block { expr: Some(block_expr), .. }, _) => {
+                self.note_type_is_not_clone_inner_expr(block_expr)
+            }
```

Comment:
similar to the `peel_blocks` use above it would be nice if `expr.peel_blocks()` is used for this match also. Then we'd avoid the unnecessary recursion and reduce the noise from handling not very relevant cases.